src/Compilers/CSharp/Portable/CodeGen/EmitArrayInitializer.cs

```diff
@@ -468,26 +524,79 @@ private bool TryEmitReadonlySpanAsBlobWrapper(NamedTypeSymbol spanType, BoundExp
+                    if (lengthForConstructor != elementCount)
+                    {
+                        // We need to use RuntimeHelpers.CreateSpan, but the code has requested a subset of the elements.
+                        // That means the code is something like `new ReadOnlySpan<char>(new[] { 'a', 'b', 'c' }, 1, 2)`
+                        // rather than `new ReadOnlySpan<char>(new[] { 'b', 'c' })`.  If such a pattern is found to be
+                        // common, this could be augmented to accomodate it.  For now, we just return false to fail
+                        // to optimize this case.
+                        return false;
+                    }
+
+                    if (inPlaceTarget is not null)
+                    {
+                        // We can use RuntimeHelpers.CreateSpan but not for in-place initialization. Fail to optimize,
+                        // but tell the caller they can call this again with a null inPlaceTarget, at which point this
+                        // should be able to optimize the call.
+                        avoidInPlace = true;
+                        return false;
+                    }
+
+                    // ldtoken <PrivateImplementationDetails>...
+                    // call ReadOnlySpan<elementType> RuntimeHelpers::CreateSpan<elementType(fldHandle)
+                    var field = _builder.module.GetFieldForData(data, (ushort)specialElementType.SizeInBytes(), wrappedExpression.Syntax, _diagnostics);
+                    _builder.EmitOpCode(ILOpCode.Ldtoken);
+                    _builder.EmitToken(field, wrappedExpression.Syntax, _diagnostics);
+                    _builder.EmitOpCode(ILOpCode.Call, stackAdjustment: 0);
+                    EmitSymbolToken(createSpan, wrappedExpression.Syntax, optArgList: null);
                 }
                 else
                 {
-                    // consumes data ptr and size, pushes the instance
-                    _builder.EmitOpCode(ILOpCode.Newobj, stackAdjustment: -1);
-                }
+                    // We're dealing with a ReadOnlySpan<byte>. Make sure it has the (void*, int) ctor.
+                    var ctor = (MethodSymbol?)_module.Compilation.GetWellKnownTypeMember(WellKnownMember.System_ReadOnlySpan_T__ctor_Pointer);
+                    if (ctor == null || ctor.GetUseSiteInfo().DiagnosticInfo?.Severity == DiagnosticSeverity.Error)
+                    {
+                        // We need to use `new ReadOnlySpan(ptr, length)`, but that constructor can't be found.
+                        return false;
+                    }
+
+                    if (inPlaceTarget is not null)
+                    {
+                        EmitAddress(inPlaceTarget, Binder.AddressKind.Writeable);
+                    }
+
+                    _builder.EmitArrayBlockFieldRef(data, (ushort)specialElementType.SizeInBytes(), wrappedExpression.Syntax, _diagnostics);
```

Comment:
I didn't understand why we're passing an alignment value here. This branch doesn't use `CreateSpan` API, and it seems like you opted to not align for scenarios that don't technically need it.

Reply:
specialElementType here is always `byte`/`sbyte`/`bool`, and thus is always 1.  I can just change it to hardcode it.

Reply:
I think we can hardcode it (ie. remove parameter from `EmitArrayBlockFieldRef`).

> specialElementType here is always byte/sbyte/bool, and thus is always 1. 
Oh, I see that now. Thanks
Consider adding an assertion in this branch.